packages/core/saved-objects/core-saved-objects-migration-server-internal/src/model/model.ts

```diff
@@ -1445,38 +1445,40 @@ export const model = (currentState: State, resW: ResponseType<AllActionStates>):
           controlState: 'UPDATE_TARGET_MAPPINGS_PROPERTIES',
           updatedTypesQuery: Option.none,
         };
-      } else if (isTypeof(left, 'root_fields_changed')) {
-        // compatible migration: some core fields have been updated
-        return {
-          ...stateP,
-          controlState: 'UPDATE_TARGET_MAPPINGS_PROPERTIES',
-          // we must "pick-up" all documents on the index (by not providing a query)
-          updatedTypesQuery: Option.none,
-          logs: [
-            ...stateP.logs,
-            {
-              level: 'info',
-              message: `Kibana is performing a compatible upgrade and the mappings of some root fields have been changed. For Elasticsearch to pickup these mappings, all saved objects need to be updated. Updated root fields: ${left.updatedFields}.`,
-            },
-          ],
-        };
-      } else if (isTypeof(left, 'types_changed')) {
-        // compatible migration: some fields have been updated, and they all correspond to SO types
-        const updatedTypesQuery = Option.fromNullable(buildPickupMappingsQuery(left.updatedTypes));
+      } else if (isTypeof(left, 'mappings_changed')) {
+        if (left.updatedTypes?.length) {
```

Comment:
nit: It's not so clean that we have do null / array length checks because `type: 'mappings_changed'` basically has two optional fields.

I think it was cleaner with the `type: 'root_fields_changed'` and explicitly handling this case, even if we say "well the root fields changed but there's nothing to do".

Alternatively we can return a `right` when the root_fields_changed to say "actually no mappings changed", since in both cases we go to `CHECK_VERSION_INDEX_READY_ACTIONS`. This perhaps makes the model slightly simpler even if it can be a little confusing "why does the action say no mappings changed if actually the root fields did change?". I don't really have a preference between these two options.

Reply:
The problem is that we have 4 combinations:
```
root_fields_changed: true,  types_changed: true
root_fields_changed: true,  types_changed: false
root_fields_changed: false, types_changed: true
root_fields_changed: false, types_changed: false
```
In previous implementation root fields had a bigger impact in the flow, so whenever `root_fields_changed` we did not care about `types_changed` (cause we picked up everything).
Now it's almost the other way round, in fact we don't care that much whether root fields change.

So having a return type saying `types_changed` feels inaccurate, as we might still have `root_fields_changed` as well. On top of that, it seems that the `diffMappings` has knowledge about what's more important for the flow (model.ts), by returning one prior to the other.

Reply:
I can either make that knowledge explicit in `diffMappings` by saying "mappings_changed_must_update_types" Vs "mappings_changed_only_root_fields" (or similar).
Or we could add booleans to prevent the `null / array length checks`, and let `model.ts` decide what to do.

Reply:
Ah I see. Yes I do agree that generally having more logic in the model makes sense, and actions are generally more "dumb" and just deal with Elasticsearch.

Part of me wants to keep root field change detection because we might need it. But we probably would never need it and it does make everything simpler if we just have `types_changed`. We won't be able to log "root fields changed" but I'm also not sure that's really necessary or relevant as it doesn't have any impact or migration performance. Do we loose anything if we completely remove root fields changed from `checkTargetMappings`?

Reply:
I think we don't lose anything besides logging, so I will simplify and rename:
```
checkTargetMappings => checkTargetTypesMappings
```

It will only contemplate the case where `types_changed`.